utils/service/txmgr/txmgr.go

```diff
@@ -89,315 +65,438 @@ type ETHBackend interface {
+	if m.cfg.TxSendTimeout != 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, m.cfg.TxSendTimeout)
+		defer cancel()
+	}
+	tx, err := m.craftTx(ctx, candidate)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create the tx: %w", err)
+	}
+	return m.send(ctx, tx)
+}
 
-	// Initialize a wait group to track any spawned goroutines, and ensure
-	// we properly clean up any dangling resources this method generates.
-	// We assert that this is the case thoroughly in our unit tests.
-	var wg sync.WaitGroup
-	defer wg.Wait()
+// craftTx creates the signed transaction
+// It queries L1 for the current fee market conditions as well as for the nonce.
+// NOTE: This method SHOULD NOT publish the resulting transaction.
+// NOTE: If the [TxCandidate.GasLimit] is non-zero, it will be used as the transaction's gas.
+// NOTE: Otherwise, the [SimpleTxManager] will query the specified backend for an estimate.
+func (m *SimpleTxManager) craftTx(ctx context.Context, candidate TxCandidate) (*types.Transaction, error) {
+	gasTipCap, basefee, err := m.suggestGasPriceCaps(ctx)
+	if err != nil {
+		m.metr.RPCError()
+		return nil, fmt.Errorf("failed to get gas price info: %w", err)
+	}
+	gasFeeCap := calcGasFeeCap(basefee, gasTipCap)
 
-	// Initialize a subcontext for the goroutines spawned in this process.
-	// The defer to cancel is done here (in reverse order of Wait) so that
-	// the goroutines can exit before blocking on the wait group.
-	ctx, cancel := context.WithCancel(ctx)
+	// Fetch the sender's nonce from the latest known block (nil `blockNumber`)
+	childCtx, cancel := context.WithTimeout(ctx, m.cfg.NetworkTimeout)
 	defer cancel()
+	nonce, err := m.backend.NonceAt(childCtx, m.cfg.From, nil)
+	if err != nil {
+		m.metr.RPCError()
+		return nil, fmt.Errorf("failed to get nonce: %w", err)
+	}
+	m.metr.RecordNonce(nonce)
 
-	sendState := NewSendState(m.SafeAbortNonceTooLowCount)
-
-	// Create a closure that will block on submitting the tx in the
-	// background, returning the first successfully mined receipt back to
-	// the main event loop via receiptChan.
-	receiptChan := make(chan *types.Receipt, 1)
-	sendTxAsync := func(tx *types.Transaction) {
-		defer wg.Done()
+	rawTx := &types.DynamicFeeTx{
+		ChainID:   m.chainID,
+		Nonce:     nonce,
+		To:        candidate.To,
+		GasTipCap: gasTipCap,
+		GasFeeCap: gasFeeCap,
+		Data:      candidate.TxData,
+	}
 
-		txHash := tx.Hash()
-		nonce := tx.Nonce()
-		gasTipCap := tx.GasTipCap()
-		gasFeeCap := tx.GasFeeCap()
-		log := m.l.New("txHash", txHash, "nonce", nonce, "gasTipCap", gasTipCap, "gasFeeCap", gasFeeCap)
-		log.Info("publishing transaction")
+	m.l.Info("creating tx", "to", rawTx.To, "from", m.cfg.From)
 
-		err := m.backend.SendTransaction(ctx, tx)
-		sendState.ProcessSendError(err)
+	// If the gas limit is set, we can use that as the gas
+	if candidate.GasLimit != 0 {
+		rawTx.Gas = candidate.GasLimit
+	} else {
+		// Calculate the intrinsic gas for the transaction
```

Comment:
Comment looks invalid. Intrinsic gas cost is a minimum gas cost required and the below tries to compute estimated gas via RPC.